ledger-core/v2/virtual/statemachine/logger.go

```diff
@@ -0,0 +1,210 @@
+		backtrace string
+		err       string
+	)
+	if stepLoggerData.Error != nil {
+		if slotPanicError, ok := stepLoggerData.Error.(smachine.SlotPanicError); ok {
+			backtrace = string(slotPanicError.Stack)
+		}
+		err = stepLoggerData.Error.Error()
+	}
+
+	if err != "" {
+		fmt.Println(err)
+	}
+
+	msg := LogStepMessage{
+		Message: special + stepLoggerUpdateData.UpdateType + suffix,
+
+		MachineName: stepLoggerData.Declaration,
+		MachineID:   fmt.Sprintf("%s[%3d]", stepLoggerData.StepNo.MachineID(), stepLoggerData.CycleNo),
+		SlotStep:    fmt.Sprintf("%03d @ %03d", stepLoggerData.StepNo.StepNo(), stepLoggerData.StepNo.StepNo()),
+
+		From: stepLoggerData.CurrentStep.GetStepName(),
+		To:   stepLoggerUpdateData.NextStep.GetStepName(),
+
+		Error:     err,
+		Backtrace: backtrace,
+	}
+
+	if stepLoggerUpdateData.ActivityNano > 0 {
+		msg.ExecutionTime = stepLoggerUpdateData.ActivityNano.Nanoseconds()
+	}
+	if stepLoggerUpdateData.InactivityNano > 0 {
+		msg.InactivityTime = stepLoggerUpdateData.InactivityNano.Nanoseconds()
+	}
+
+	c.logger.Error(msg)
+}
+
+type ConveyorLoggerFactory struct {
+}
+
+func (c ConveyorLoggerFactory) CreateStepLogger(ctx context.Context, _ smachine.StateMachine, traceID smachine.TracerID) smachine.StepLogger {
+	_, logger := inslogger.WithTraceField(context.Background(), traceID)
+	return &ConveyorLogger{
+		StepLoggerStub: smachine.StepLoggerStub{TracerID: traceID},
+		logger:         logger,
+	}
+}
+
+type LogInternal struct {
+	*log.Msg `txt:"internal"`
+
+	Message   string `fmt:"internal - %s"`
+	Component string `txt:"sm"`
+
+	MachineID string
+	SlotStep  string
+	Error     error  `opt:""`
+	Backtrace string `opt:""`
+}
+
+func (ConveyorLoggerFactory) LogMachineInternal(slotMachineData smachine.SlotMachineData, msg string) {
+	backtrace := ""
+	if slotMachineData.Error != nil {
+		if slotPanicError, ok := slotMachineData.Error.(smachine.SlotPanicError); ok {
+			backtrace = string(slotPanicError.Stack)
+		}
+	}
+	global.Logger().Error(LogInternal{
+		Message: msg,
+
+		MachineID: fmt.Sprintf("%s[%3d]", slotMachineData.StepNo.MachineID(), slotMachineData.CycleNo),
+		SlotStep:  fmt.Sprintf("%03d @ %03d", slotMachineData.StepNo.StepNo(), slotMachineData.StepNo.StepNo()),
+		Error:     slotMachineData.Error,
+		Backtrace: backtrace,
+	})
+}
+
+type LogCritical struct {
+	*log.Msg `txt:"internal"`
+
+	Message   string `fmt:"internal critical - %s"`
+	Component string `txt:"sm"`
+
+	MachineID string
+	SlotStep  string
+	Error     error  `opt:""`
+	Backtrace string `opt:""`
+}
+
+func (ConveyorLoggerFactory) LogMachineCritical(slotMachineData smachine.SlotMachineData, msg string) {
+	backtrace := ""
+	if slotMachineData.Error != nil {
+		if slotPanicError, ok := slotMachineData.Error.(smachine.SlotPanicError); ok {
+			backtrace = string(slotPanicError.Stack)
+		}
+	}
+	global.Logger().Error(LogCritical{
+		Message: msg,
+
+		MachineID: fmt.Sprintf("%s[%3d]", slotMachineData.StepNo.MachineID(), slotMachineData.CycleNo),
```

Comment:
Please avoid formating in logging. Its better to have separate field CycleNo and StepNo